python/rpdk/java/templates/init/guided_aws/TagHelper.java

```diff
@@ -0,0 +1,250 @@
+     * @return Converted Map of tags
+     */
+    public static Map<String, String> convertToMap(final Collection<Tag> tags) {
+        if (CollectionUtils.isEmpty(tags)) {
+            return Collections.emptyMap();
+        }
+        return tags.stream()
+            .filter(tag -> tag.value() != null)
+            .collect(Collectors.toMap(
+                Tag::key,
+                Tag::value,
+                (oldValue, newValue) -> newValue));
+    }
+
+    /**
+     * convertToSet
+     *
+     * Converts a tag map to a set of Tag objects.
+     *
+     * Note: Like convertToMap, convertToSet filters out value-less tag entries.
+     *
+     * @param tagMap Map of tags to convert
+     * @return Set of Tag objects
+     */
+    public static Set<Tag> convertToSet(final Map<String, String> tagMap) {
+        if (MapUtils.isEmpty(tagMap)) {
+            return Collections.emptySet();
+        }
+        return tagMap.entrySet().stream()
+            .filter(tag -> tag.getValue() != null)
+            .map(tag -> Tag.builder()
+                .key(tag.getKey())
+                .value(tag.getValue())
+                .build())
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * shouldUpdateTags
+     *
+     * Generate tags to put into resource creation request.
+     * This includes user defined tags and system tags as well.
+     */
+    public final Map<String, String> generateTagsForCreate(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> tagMap = new HashMap<>();
+
+        if (handlerRequest.getSystemTags() != null) {
+            tagMap.putAll(handlerRequest.getSystemTags());
+        }
+
+        if (handlerRequest.getDesiredResourceTags() != null) {
+            tagMap.putAll(handlerRequest.getDesiredResourceTags());
+        }
+
+        // TODO: get tags from resource model based on your tag property name
+        // TODO: tagMap.putAll(convertToMap(resourceModel.getTags()));
+        return Collections.unmodifiableMap(tagMap);
+    }
+
+    /**
+     * shouldUpdateTags
+     *
+     * Determines whether user defined tags or system tags have been changed during update.
+     */
+    public final boolean shouldUpdateTags(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        return areResourceTagsChanging(resourceModel, handlerRequest) || areSystemTagsChanging(handlerRequest);
+    }
+
+    private boolean areResourceTagsChanging(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> currentTags = getCurrentTags(handlerRequest);
+        final Map<String, String> desiredTags = getNewDesiredTags(resourceModel, handlerRequest);
+        return ObjectUtils.notEqual(currentTags, desiredTags);
+    }
+
+    private Map<String, String> getCurrentTags(final ResourceHandlerRequest<ResourceModel> handlerRequest) {
```

Comment:
current is confusing. maybe `getPreviouslyAttachedTags`? why not including stack-level tags?

Reply:
stack-level tags are included, if service team didn't override their config file, handlerRequest.getPreviousResourceTags() is previous stack level tags and handlerRequest.getPreviousResourceState().getTags() are resource tags. Correct if I'm wrong about it. But sure I can rename the function or adding more annotations.